### ISSUE TITLE
Make PreProcessors work like Middleware so then can perform a before/after

### ISSUE THREAD

**Issue by Shazwazza** [Shazwazza/Smidge#53]
It would be nice if pre-processors worked like Middleware so they could call the next execution pipeline and then execute some logic afterwords too.

This would make it much more simple to do things like css source maps, and probably could come in handy for various other things too.

**Comment by Shazwazza** [comment]
Have pushed a branch for this.

@dazinator wondering if you think this is a good idea or if it does/doesn't add any benefit. Would mean that a single pre-processor for a single file could run before and after another pre-processor (i.e. just like Middelware)

**Comment by dazinator** [comment]
Interesting concept! 
My take on this is that the main benefit to this pattern is it will allow the processors to optionally:

1. Short circuit / block all further processing.
2. Conditionally skip processing and pass the file content on to the next processor untouched.
3. Append some meta data to the file before the next pricessor then remove it after.

As to how / when and why a processor would decide to do any of the above, i'm not too sure, and without having some use cases it's hard to see the real benefits..

I'll try and contrive a few.

I don't think number 1 is really safe for a file processor to do. So I can't see a benefit their. It makes sense for middleware to block http requests and generate a response. For a file processing requirements I could only see this yielding incomplete files. One processor doesn't really ever know if it's safe or not to not call the next one - because if it gets it wrong then you won't see the file output you expect.

For number 2. Perhaps if you had a processor that minifies. If you passed some options to it to say don't minify in debug mode, then it could skip processing and pass the input through to the next processors. Otherwise when not in debug mode it would process as normal. What would the next processor be though to make this workflow conpeoling? Not sure. And can't we already do this - as currently a processor would just return the input content as the result.

For number 3.
This could be useful.. Im imagining some kind of source map inlining processor. It starts by pass ing the file to the next processor then after that has run it checks the file content for a source map declaration. If it finds a declaration to an external .map file, it could go read that map file then inline the sourcemap into the file content itself instead. This means for processors that produce seperate sourcemap files, smidge wouldn't have to worry about placing those source map files somewhere they can be served up to the browser - as it could just inline the source map instead.

However give that most processors worth their salt will often have options for producing inline source maps.. not really sure how useful this would really be.

Adding a header / footer to every file?

Run out of ideas!

**Comment by Shazwazza** [comment]
Yeah those were basically my thoughts too - i thought it could be a neat/fun idea (which it was) but can't quite think of a real compelling reason to put it into the release apart from just being extra flexible. Header/Footer is certainly an option if someone needed that. Will keep this here for now and think on it a bit more :)
